src/primitives.rs

```diff
@@ -0,0 +1,183 @@
+    #[serde(rename = "minerAddr")]
+    miner_addr: String,
+    #[serde(rename = "extraData")]
+    extra_data: String,
+    transactions: Vec<String>,
+    #[serde(rename = "merkleHashes")]
+    merkle_hashes: Vec<String>,
+    #[serde(rename = "prunedAccounts")]
+    pruned_accounts: Vec<String>,
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct GetWork {
+	pub data: String,
+	pub suffix: String,
+	pub target: u64,
+	pub algorithm: String
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct PeerList {
+	pub id: String,
+	pub address: String,
+	#[serde(rename = "addressState")]
+	pub address_state: i64,
+	#[serde(rename = "connectionState")]
+	pub connection_state: Option<i64>,
+	pub version: Option<i64>,
+	#[serde(rename = "timeOffset")]
+	pub time_offset: Option<i64>,
+	#[serde(rename = "headHash")]
+	pub head_hash: Option<String>,
+	pub latency: Option<i64>,
+	pub rx: Option<i64>,
+	pub tx: Option<i64>
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct PeerState {
+	pub id: String,
+	pub address: String,
+	#[serde(rename = "addressState")]
+	pub address_state: u8
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub enum Syncing {
+	IsSyncing(bool),
+	Pending(Pending)
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct Pending {
+	#[serde(rename = "startingBlock")]
+	starting_block: u64,
+	#[serde(rename = "currentBlock")]
+	current_block: u64,
+	#[serde(rename = "highestBlock")]
+	highest_block: u64
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct Transaction {
+	pub hash: String,
+	#[serde(rename = "blockHash")]
+	pub block_hash: String,
+	#[serde(rename = "blockNumber")]
+	pub block_number: u64,
+	pub timestamp: u64,
+	pub confirmations: u64,
+	#[serde(rename = "transactionIndex")]
+	#[serde(skip_deserializing)] //TODO skip deserializing, have to look into making field optional
+	transaction_index: ::serde_json::Value,
+	pub from: String,
+	#[serde(rename = "fromAddress")]
+	pub from_address: String,
+	pub to: String,
+	#[serde(rename = "toAddress")]
+	pub to_address: String,
+	pub value: u64,
+	pub fee: u64,
+	pub data: ::serde_json::Value,
+	pub flags: u32
+}
+
+#[derive(Debug, serde_derive::Serialize, serde_derive::Deserialize)]
+pub struct TransactionReceipt {
+	#[serde(rename = "transactionHash")]
+	pub transaction_hash: String,
+	#[serde(rename = "transactionIndex")]
+	pub transaction_index: i64,
+	#[serde(rename = "blockNumber")]
+	pub block_number: i64,
+	#[serde(rename = "blockHash")]
+	pub block_hash: String,
+	pub confirmations: i64,
+	pub timestamp: i64
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(untagged)]
+pub enum TransactionSequence{
+	BlockHashes(Vec<String>),
+	Transactions(Vec<Transaction>)
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct OutgoingTransaction {
+	pub from: &'static str,
```

Comment:
These should probably better be `String` as you can't assume these variables to have a static lifetime.

Reply:
If I would change them to String, you have to add `to_string()` to both `to` and `from`

```rust
let tx = primitives::OutgoingTransaction {
		   from: "NQ32 R6DB VFM5 M931 7X4E 0N5Q LJ56 9QCR 4T42".to_string(),
		   to: "NQ74 61S8 2FD3 RVPG HU09 1Y57 77E6 BL38 TQH3".to_string(),
		   value: 100, //Lunas
		   fee: 0
		};
```

Because it expects a struct `std::string::String` but otherwise receives a reference.

Reply:
I think the `.to_string()` here is fine.
The problem with `&'static str` is that it only accepts string literals, i.e., something like `"abc"`. If I get the address from user input via the terminal or a GUI, for example, it will not have the required `'static` lifetime.

So, either you need to put a lifetime parameter in there:
```
pub struct OutgoingTransaction<'a> {
    pub from: &'a str,
    ...
}
```
or replace it by `String`.

The lifetime parameter option will work right now. However, if you ever want to have a function that returns a `OutgoingTransaction` object that was received via the JSON Rpc, you'll run into trouble. That is because you received these strings via the network and somebody needs to be the owner of the strings when they are returned.

Reply:
I changed the `from` and `to` types to `String`